src/main/java/org/terasology/metalrenegades/interaction/systems/WellWaterSystem.java

```diff
@@ -15,36 +15,178 @@
 import org.terasology.metalrenegades.interaction.component.WellSourceComponent;
+import org.terasology.metalrenegades.interaction.events.CupFilledEvent;
+import org.terasology.metalrenegades.interaction.events.WellDrinkEvent;
+import org.terasology.metalrenegades.interaction.events.WellRefilledEvent;
 import org.terasology.registry.In;
+import org.terasology.thirst.component.ThirstComponent;
+import org.terasology.thirst.event.DrinkConsumedEvent;
+import org.terasology.world.time.WorldTimeEvent;
+
+import java.util.stream.StreamSupport;
 
 /**
- * Tracks water source blocks inside of wells; and gives water cups to players upon interaction.
+ * Tracks water source blocks inside of wells, fills player's water cups upon interaction, and empties
+ * water cups upon consumption.
  */
 @RegisterSystem(RegisterMode.AUTHORITY)
 public class WellWaterSystem extends BaseComponentSystem {
 
     @In
     private EntityManager entityManager;
 
-    @ReceiveEvent(components = {WellSourceComponent.class})
-    public void onActivate(ActivateEvent event, EntityRef target) {
+    @In
+    private Time time;
+
+    /**
+     * The number of world time cycles it takes for all wells to replenish with one refill.
+     */
+    private static final int CYCLES_UNTIL_REFILL = 20;
+
+    private static final String EMPTY_CUP_URI = "MetalRenegades:emptyCup";
+
+    private static final String FULL_CUP_URI = "MetalRenegades:filledCup";
+
+    /**
+     * A timer counting the current number of cycles until all wells are replenished.
+     */
+    private int worldTimeCycles = 0;
+
+    @ReceiveEvent
+    public void onWorldTimeEvent(WorldTimeEvent worldTimeEvent, EntityRef entityRef) {
+        if (worldTimeCycles >= CYCLES_UNTIL_REFILL) {
+            refillWells();
+            worldTimeCycles = 0;
+        }
+
+        worldTimeCycles++;
+    }
+
+    @ReceiveEvent(components = {WellBlockComponent.class})
+    public void onActivate(ActivateEvent event, EntityRef sourceBlock) {
         EntityRef gatheringCharacter = event.getInstigator();
-        EntityRef cupItem = entityManager.create("MetalRenegades:waterCup");
+        EntityRef wellEntity = getWellEntity(sourceBlock);
+        if (!gatheringCharacter.exists() || wellEntity == EntityRef.NULL) {
+            return;
+        }
 
-        if(!cupItem.exists() || !gatheringCharacter.exists()) {
+        WellSourceComponent wellSourceComp = wellEntity.getComponent(WellSourceComponent.class);
+        if (!wellSourceComp.useRefill()) { // if no refills remain, don't give water.
             return;
         }
 
+        EntityRef heldItem = gatheringCharacter.getComponent(CharacterHeldItemComponent.class).selectedItem;
+        if (heldItem.hasComponent(WaterCupComponent.class)) {
+            fillWaterCup(gatheringCharacter, heldItem, wellEntity);
+        } else {
+            directDrink(gatheringCharacter, wellEntity);
+        }
+
+        wellEntity.saveComponent(wellSourceComp);
```

Comment:
Maybe you can tweak the method flow in a way that the first part is to retrieve the amount of water that can be taken from the well (`gatherWater`) and the second part doing something with this amount (`consumeWater`). 
In particular, the second method should be able to deal with a zero-amount, thus getting rid of the addition `return` in here altogether.

```suggestion
		// take out water if possible and save the updated component
        int fillingAmount = gatherWater(wellEntity);
		consumeWater(fillingAmount, gatheringCharacter);
```

Reply:
I didn't use exactly what the code suggestion says, but something similar that still removes the additional return statement 👍